### PR TITLE
Enhancements

### PR DESCRIPTION
**Pull request by epmoyer**
William, 

Thanks for the excellent work refactoring to_precision! I'd been working on my own refactor of Randle's project but somehow I'd missed yours, so now I decided to scrap mine and merge the changes with your code base instead.
 
### Fixes:
* 0 should not return e notation in 'auto' mode.
    * Was `to_pricision(0, 1) => '0.e0'`, is now `to_pricision(0, 1) => '0'`
* Should not return numbers with trailing decimals 
    * Was `to_precision(0, 1, 'std') => '0.'`, is now `to_precision(0, 1, 'std') => '0'`
    * Eliminated trailing "." for all results
    * Corrected associated test cases
* Fixed comments in _place_dot; added quotes around digits argument in examples, and around results
* Added comments for `filler` option
* Fixed heading level of "scientific notation" in README
* Minor spelling fixes in code and README

### Additions:
* Added option `strip_zeros`
* Added option `auto_limit`
* Added option `preserve`
* Added test cases for the new options
* Added "Implicit vs. Explicit" section to README
* Added `demonstration.py`
* Added "Demonstration" section to README (output of `demonstration.py`)

The current version of `demonstration.py` writes out a JSON file containing test vectors for all the precisions & options it exercises.  On my (scrapped) project I had been using those vectors to drive my validation testing.  The downside is that one has to carefully inspect the output of `demonstration.py` to make sure that the vectors are correct, but the upside is that once you lock in the test vectors you then have a very large coverage set (2280 vectors today) to run future changes against.  A nice compromise would be to run both the existing hand written tests and check the vectors; that way there is very broad test coverage without giving up confidence in the current set.  Anyway, I have the vector validation code lying around and can port it over if you are interested in adding the vectors to the test suite; it's ~10 lines of code in py.test today.

    import json
    import pytest

    from precision.precision import to_precision

    with open('test_vectors.json') as data_file:
        test_vectors_as_dicts = json.load(data_file)
    test_vectors = [(v['input_parameters'], v['expected_output']) for v in test_vectors_as_dicts]

    @pytest.mark.parametrize('input_parameters, expected_output', test_vectors)
    def test_precision(input_parameters, expected_output):
        result_string = to_precision(**input_parameters)
        assert result_string == expected_output

The `demonstration.py` module can also be used on the previous version of `to_precision.py` (to compare its behavior) using the `--legacy` option (which suppresses use of the new features added in my branch)

The results running against the legacy version are below.  The results running against my branch are in the updated README.

Cheers!

    $./demonstration.py --legacy
    value         precision=1   precision=2    precision=3     precision=4      precision=5       Additional Options  
    ------------  ------------  -------------  --------------  ---------------  ----------------  ------------------  
    0             0.e0          0.0e0          0.00e0          0.000e0          0.0000e0                              
    10            10            10.            10.0            10.00            10.000                                
    100           100           100            100.            100.0            100.00                                
    1             1             1.0            1.00            1.000            1.0000                                
    12            10            12             12.0            12.00            12.000                                
    123           100           120            123             123.0            123.00                                
    1234          1e3           1.2e3          1.23e3          1.234e3          1.2340e3                              
    12345         1e4           1.2e4          1.23e4          1.234e4          1.2345e4                              
    123456        1e5           1.2e5          1.23e5          1.235e5          1.2346e5                              
    1234567       1e6           1.2e6          1.23e6          1.235e6          1.2346e6                              
    12345678      1e7           1.2e7          1.23e7          1.235e7          1.2346e7                              
    123456789     1e8           1.2e8          1.23e8          1.235e8          1.2346e8                              
    1.2           1             1.2            1.20            1.200            1.2000                                
    1.23          1             1.2            1.23            1.230            1.2300                                
    1.234         1             1.2            1.23            1.234            1.2340                                
    1.2345        1             1.2            1.23            1.234            1.2345                                
    1.23456       1             1.2            1.23            1.235            1.2346                                
    1.234567      1             1.2            1.23            1.235            1.2346                                
    1.2345678     1             1.2            1.23            1.235            1.2346                                
    1.23456789    1             1.2            1.23            1.235            1.2346                                
    0.000000009   9e-9          9.0e-9         9.00e-9         9.000e-9         9.0000e-9                             
    0.000000089   9e-8          8.9e-8         8.90e-8         8.900e-8         8.9000e-8                             
    0.000000789   8e-7          7.9e-7         7.89e-7         7.890e-7         7.8900e-7                             
    0.000006789   7e-6          6.8e-6         6.79e-6         6.789e-6         6.7890e-6                             
    0.000056789   6e-5          5.7e-5         5.68e-5         5.679e-5         5.6789e-5                             
    0.000456789   5e-4          4.6e-4         4.57e-4         4.568e-4         4.5679e-4                             
    0.003456789   0.003         0.0035         0.00346         0.003457         0.0034568                             
    0.023456789   0.02          0.023          0.0235          0.02346          0.023457                              
    0.123456789   0.1           0.12           0.123           0.1235           0.12346                               
    -10           -10           -10.           -10.0           -10.00           -10.000                               
    -100          -100          -100           -100.           -100.0           -100.00                               
    -1            -1            -1.0           -1.00           -1.000           -1.0000                               
    -12           -10           -12            -12.0           -12.00           -12.000                               
    -123          -100          -120           -123            -123.0           -123.00                               
    -1234         -1e3          -1.2e3         -1.23e3         -1.234e3         -1.2340e3                             
    -12345        -1e4          -1.2e4         -1.23e4         -1.234e4         -1.2345e4                             
    -123456       -1e5          -1.2e5         -1.23e5         -1.235e5         -1.2346e5                             
    -1234567      -1e6          -1.2e6         -1.23e6         -1.235e6         -1.2346e6                             
    -12345678     -1e7          -1.2e7         -1.23e7         -1.235e7         -1.2346e7                             
    -123456789    -1e8          -1.2e8         -1.23e8         -1.235e8         -1.2346e8                             
    -1.2          -1            -1.2           -1.20           -1.200           -1.2000                               
    -1.23         -1            -1.2           -1.23           -1.230           -1.2300                               
    -1.234        -1            -1.2           -1.23           -1.234           -1.2340                               
    -1.2345       -1            -1.2           -1.23           -1.234           -1.2345                               
    -1.23456      -1            -1.2           -1.23           -1.235           -1.2346                               
    -1.234567     -1            -1.2           -1.23           -1.235           -1.2346                               
    -1.2345678    -1            -1.2           -1.23           -1.235           -1.2346                               
    -1.23456789   -1            -1.2           -1.23           -1.235           -1.2346                               
    -0.000000009  -9e-9         -9.0e-9        -9.00e-9        -9.000e-9        -9.0000e-9                            
    -0.000000089  -9e-8         -8.9e-8        -8.90e-8        -8.900e-8        -8.9000e-8                            
    -0.000000789  -8e-7         -7.9e-7        -7.89e-7        -7.890e-7        -7.8900e-7                            
    -0.000006789  -7e-6         -6.8e-6        -6.79e-6        -6.789e-6        -6.7890e-6                            
    -0.000056789  -6e-5         -5.7e-5        -5.68e-5        -5.679e-5        -5.6789e-5                            
    -0.000456789  -5e-4         -4.6e-4        -4.57e-4        -4.568e-4        -4.5679e-4                            
    -0.003456789  -0.003        -0.0035        -0.00346        -0.003457        -0.0034568                            
    -0.023456789  -0.02         -0.023         -0.0235         -0.02346         -0.023457                             
    -0.123456789  -0.1          -0.12          -0.123          -0.1235          -0.12346                              
                                                                                                                     
    0             0.            0.0            0.00            0.000            0.0000            notation=std        
    10            10            10.            10.0            10.00            10.000            notation=std        
    100           100           100            100.            100.0            100.00            notation=std        
    1             1             1.0            1.00            1.000            1.0000            notation=std        
    12            10            12             12.0            12.00            12.000            notation=std        
    123           100           120            123             123.0            123.00            notation=std        
    1234          1000          1200           1230            1234             1234.0            notation=std        
    12345         10000         12000          12300           12340            12345             notation=std        
    123456        100000        120000         123000          123500           123460            notation=std        
    1234567       1000000       1200000        1230000         1235000          1234600           notation=std        
    12345678      10000000      12000000       12300000        12350000         12346000          notation=std        
    123456789     100000000     120000000      123000000       123500000        123460000         notation=std        
    1.2           1             1.2            1.20            1.200            1.2000            notation=std        
    1.23          1             1.2            1.23            1.230            1.2300            notation=std        
    1.234         1             1.2            1.23            1.234            1.2340            notation=std        
    1.2345        1             1.2            1.23            1.234            1.2345            notation=std        
    1.23456       1             1.2            1.23            1.235            1.2346            notation=std        
    1.234567      1             1.2            1.23            1.235            1.2346            notation=std        
    1.2345678     1             1.2            1.23            1.235            1.2346            notation=std        
    1.23456789    1             1.2            1.23            1.235            1.2346            notation=std        
    0.000000009   0.000000009   0.0000000090   0.00000000900   0.000000009000   0.0000000090000   notation=std        
    0.000000089   0.00000009    0.000000089    0.0000000890    0.00000008900    0.000000089000    notation=std        
    0.000000789   0.0000008     0.00000079     0.000000789     0.0000007890     0.00000078900     notation=std        
    0.000006789   0.000007      0.0000068      0.00000679      0.000006789      0.0000067890      notation=std        
    0.000056789   0.00006       0.000057       0.0000568       0.00005679       0.000056789       notation=std        
    0.000456789   0.0005        0.00046        0.000457        0.0004568        0.00045679        notation=std        
    0.003456789   0.003         0.0035         0.00346         0.003457         0.0034568         notation=std        
    0.023456789   0.02          0.023          0.0235          0.02346          0.023457          notation=std        
    0.123456789   0.1           0.12           0.123           0.1235           0.12346           notation=std        
    -10           -10           -10.           -10.0           -10.00           -10.000           notation=std        
    -100          -100          -100           -100.           -100.0           -100.00           notation=std        
    -1            -1            -1.0           -1.00           -1.000           -1.0000           notation=std        
    -12           -10           -12            -12.0           -12.00           -12.000           notation=std        
    -123          -100          -120           -123            -123.0           -123.00           notation=std        
    -1234         -1000         -1200          -1230           -1234            -1234.0           notation=std        
    -12345        -10000        -12000         -12300          -12340           -12345            notation=std        
    -123456       -100000       -120000        -123000         -123500          -123460           notation=std        
    -1234567      -1000000      -1200000       -1230000        -1235000         -1234600          notation=std        
    -12345678     -10000000     -12000000      -12300000       -12350000        -12346000         notation=std        
    -123456789    -100000000    -120000000     -123000000      -123500000       -123460000        notation=std        
    -1.2          -1            -1.2           -1.20           -1.200           -1.2000           notation=std        
    -1.23         -1            -1.2           -1.23           -1.230           -1.2300           notation=std        
    -1.234        -1            -1.2           -1.23           -1.234           -1.2340           notation=std        
    -1.2345       -1            -1.2           -1.23           -1.234           -1.2345           notation=std        
    -1.23456      -1            -1.2           -1.23           -1.235           -1.2346           notation=std        
    -1.234567     -1            -1.2           -1.23           -1.235           -1.2346           notation=std        
    -1.2345678    -1            -1.2           -1.23           -1.235           -1.2346           notation=std        
    -1.23456789   -1            -1.2           -1.23           -1.235           -1.2346           notation=std        
    -0.000000009  -0.000000009  -0.0000000090  -0.00000000900  -0.000000009000  -0.0000000090000  notation=std        
    -0.000000089  -0.00000009   -0.000000089   -0.0000000890   -0.00000008900   -0.000000089000   notation=std        
    -0.000000789  -0.0000008    -0.00000079    -0.000000789    -0.0000007890    -0.00000078900    notation=std        
    -0.000006789  -0.000007     -0.0000068     -0.00000679     -0.000006789     -0.0000067890     notation=std        
    -0.000056789  -0.00006      -0.000057      -0.0000568      -0.00005679      -0.000056789      notation=std        
    -0.000456789  -0.0005       -0.00046       -0.000457       -0.0004568       -0.00045679       notation=std        
    -0.003456789  -0.003        -0.0035        -0.00346        -0.003457        -0.0034568        notation=std        
    -0.023456789  -0.02         -0.023         -0.0235         -0.02346         -0.023457         notation=std        
    -0.123456789  -0.1          -0.12          -0.123          -0.1235          -0.12346          notation=std        
                                                                                                                     
    0             0.e0          0.0e0          0.00e0          0.000e0          0.0000e0          notation=eng        
    10            10e0          10.e0          10.0e0          10.00e0          10.000e0          notation=eng        
    100           100e0         100e0          100.e0          100.0e0          100.00e0          notation=eng        
    1             1e0           1.0e0          1.00e0          1.000e0          1.0000e0          notation=eng        
    12            10e0          12e0           12.0e0          12.00e0          12.000e0          notation=eng        
    123           100e0         120e0          123e0           123.0e0          123.00e0          notation=eng        
    1234          1e3           1.2e3          1.23e3          1.234e3          1.2340e3          notation=eng        
    12345         10e3          12e3           12.3e3          12.34e3          12.345e3          notation=eng        
    123456        100e3         120e3          123e3           123.5e3          123.46e3          notation=eng        
    1234567       1e6           1.2e6          1.23e6          1.235e6          1.2346e6          notation=eng        
    12345678      10e6          12e6           12.3e6          12.35e6          12.346e6          notation=eng        
    123456789     100e6         120e6          123e6           123.5e6          123.46e6          notation=eng        
    1.2           1e0           1.2e0          1.20e0          1.200e0          1.2000e0          notation=eng        
    1.23          1e0           1.2e0          1.23e0          1.230e0          1.2300e0          notation=eng        
    1.234         1e0           1.2e0          1.23e0          1.234e0          1.2340e0          notation=eng        
    1.2345        1e0           1.2e0          1.23e0          1.234e0          1.2345e0          notation=eng        
    1.23456       1e0           1.2e0          1.23e0          1.235e0          1.2346e0          notation=eng        
    1.234567      1e0           1.2e0          1.23e0          1.235e0          1.2346e0          notation=eng        
    1.2345678     1e0           1.2e0          1.23e0          1.235e0          1.2346e0          notation=eng        
    1.23456789    1e0           1.2e0          1.23e0          1.235e0          1.2346e0          notation=eng        
    0.000000009   9e-9          9.0e-9         9.00e-9         9.000e-9         9.0000e-9         notation=eng        
    0.000000089   90e-9         89e-9          89.0e-9         89.00e-9         89.000e-9         notation=eng        
    0.000000789   800e-9        790e-9         789e-9          789.0e-9         789.00e-9         notation=eng        
    0.000006789   7e-6          6.8e-6         6.79e-6         6.789e-6         6.7890e-6         notation=eng        
    0.000056789   60e-6         57e-6          56.8e-6         56.79e-6         56.789e-6         notation=eng        
    0.000456789   500e-6        460e-6         457e-6          456.8e-6         456.79e-6         notation=eng        
    0.003456789   3e-3          3.5e-3         3.46e-3         3.457e-3         3.4568e-3         notation=eng        
    0.023456789   20e-3         23e-3          23.5e-3         23.46e-3         23.457e-3         notation=eng        
    0.123456789   100e-3        120e-3         123e-3          123.5e-3         123.46e-3         notation=eng        
    -10           -10e0         -10.e0         -10.0e0         -10.00e0         -10.000e0         notation=eng        
    -100          -100e0        -100e0         -100.e0         -100.0e0         -100.00e0         notation=eng        
    -1            -1e0          -1.0e0         -1.00e0         -1.000e0         -1.0000e0         notation=eng        
    -12           -10e0         -12e0          -12.0e0         -12.00e0         -12.000e0         notation=eng        
    -123          -100e0        -120e0         -123e0          -123.0e0         -123.00e0         notation=eng        
    -1234         -1e3          -1.2e3         -1.23e3         -1.234e3         -1.2340e3         notation=eng        
    -12345        -10e3         -12e3          -12.3e3         -12.34e3         -12.345e3         notation=eng        
    -123456       -100e3        -120e3         -123e3          -123.5e3         -123.46e3         notation=eng        
    -1234567      -1e6          -1.2e6         -1.23e6         -1.235e6         -1.2346e6         notation=eng        
    -12345678     -10e6         -12e6          -12.3e6         -12.35e6         -12.346e6         notation=eng        
    -123456789    -100e6        -120e6         -123e6          -123.5e6         -123.46e6         notation=eng        
    -1.2          -1e0          -1.2e0         -1.20e0         -1.200e0         -1.2000e0         notation=eng        
    -1.23         -1e0          -1.2e0         -1.23e0         -1.230e0         -1.2300e0         notation=eng        
    -1.234        -1e0          -1.2e0         -1.23e0         -1.234e0         -1.2340e0         notation=eng        
    -1.2345       -1e0          -1.2e0         -1.23e0         -1.234e0         -1.2345e0         notation=eng        
    -1.23456      -1e0          -1.2e0         -1.23e0         -1.235e0         -1.2346e0         notation=eng        
    -1.234567     -1e0          -1.2e0         -1.23e0         -1.235e0         -1.2346e0         notation=eng        
    -1.2345678    -1e0          -1.2e0         -1.23e0         -1.235e0         -1.2346e0         notation=eng        
    -1.23456789   -1e0          -1.2e0         -1.23e0         -1.235e0         -1.2346e0         notation=eng        
    -0.000000009  -9e-9         -9.0e-9        -9.00e-9        -9.000e-9        -9.0000e-9        notation=eng        
    -0.000000089  -90e-9        -89e-9         -89.0e-9        -89.00e-9        -89.000e-9        notation=eng        
    -0.000000789  -800e-9       -790e-9        -789e-9         -789.0e-9        -789.00e-9        notation=eng        
    -0.000006789  -7e-6         -6.8e-6        -6.79e-6        -6.789e-6        -6.7890e-6        notation=eng        
    -0.000056789  -60e-6        -57e-6         -56.8e-6        -56.79e-6        -56.789e-6        notation=eng        
    -0.000456789  -500e-6       -460e-6        -457e-6         -456.8e-6        -456.79e-6        notation=eng        
    -0.003456789  -3e-3         -3.5e-3        -3.46e-3        -3.457e-3        -3.4568e-3        notation=eng        
    -0.023456789  -20e-3        -23e-3         -23.5e-3        -23.46e-3        -23.457e-3        notation=eng        
    -0.123456789  -100e-3       -120e-3        -123e-3         -123.5e-3        -123.46e-3        notation=eng        
                                                                                                                     
    0             0.e0          0.0e0          0.00e0          0.000e0          0.0000e0          notation=sci        
    10            1e1           1.0e1          1.00e1          1.000e1          1.0000e1          notation=sci        
    100           1e2           1.0e2          1.00e2          1.000e2          1.0000e2          notation=sci        
    1             1e0           1.0e0          1.00e0          1.000e0          1.0000e0          notation=sci        
    12            1e1           1.2e1          1.20e1          1.200e1          1.2000e1          notation=sci        
    123           1e2           1.2e2          1.23e2          1.230e2          1.2300e2          notation=sci        
    1234          1e3           1.2e3          1.23e3          1.234e3          1.2340e3          notation=sci        
    12345         1e4           1.2e4          1.23e4          1.234e4          1.2345e4          notation=sci        
    123456        1e5           1.2e5          1.23e5          1.235e5          1.2346e5          notation=sci        
    1234567       1e6           1.2e6          1.23e6          1.235e6          1.2346e6          notation=sci        
    12345678      1e7           1.2e7          1.23e7          1.235e7          1.2346e7          notation=sci        
    123456789     1e8           1.2e8          1.23e8          1.235e8          1.2346e8          notation=sci        
    1.2           1e0           1.2e0          1.20e0          1.200e0          1.2000e0          notation=sci        
    1.23          1e0           1.2e0          1.23e0          1.230e0          1.2300e0          notation=sci        
    1.234         1e0           1.2e0          1.23e0          1.234e0          1.2340e0          notation=sci        
    1.2345        1e0           1.2e0          1.23e0          1.234e0          1.2345e0          notation=sci        
    1.23456       1e0           1.2e0          1.23e0          1.235e0          1.2346e0          notation=sci        
    1.234567      1e0           1.2e0          1.23e0          1.235e0          1.2346e0          notation=sci        
    1.2345678     1e0           1.2e0          1.23e0          1.235e0          1.2346e0          notation=sci        
    1.23456789    1e0           1.2e0          1.23e0          1.235e0          1.2346e0          notation=sci        
    0.000000009   9e-9          9.0e-9         9.00e-9         9.000e-9         9.0000e-9         notation=sci        
    0.000000089   9e-8          8.9e-8         8.90e-8         8.900e-8         8.9000e-8         notation=sci        
    0.000000789   8e-7          7.9e-7         7.89e-7         7.890e-7         7.8900e-7         notation=sci        
    0.000006789   7e-6          6.8e-6         6.79e-6         6.789e-6         6.7890e-6         notation=sci        
    0.000056789   6e-5          5.7e-5         5.68e-5         5.679e-5         5.6789e-5         notation=sci        
    0.000456789   5e-4          4.6e-4         4.57e-4         4.568e-4         4.5679e-4         notation=sci        
    0.003456789   3e-3          3.5e-3         3.46e-3         3.457e-3         3.4568e-3         notation=sci        
    0.023456789   2e-2          2.3e-2         2.35e-2         2.346e-2         2.3457e-2         notation=sci        
    0.123456789   1e-1          1.2e-1         1.23e-1         1.235e-1         1.2346e-1         notation=sci        
    -10           -1e1          -1.0e1         -1.00e1         -1.000e1         -1.0000e1         notation=sci        
    -100          -1e2          -1.0e2         -1.00e2         -1.000e2         -1.0000e2         notation=sci        
    -1            -1e0          -1.0e0         -1.00e0         -1.000e0         -1.0000e0         notation=sci        
    -12           -1e1          -1.2e1         -1.20e1         -1.200e1         -1.2000e1         notation=sci        
    -123          -1e2          -1.2e2         -1.23e2         -1.230e2         -1.2300e2         notation=sci        
    -1234         -1e3          -1.2e3         -1.23e3         -1.234e3         -1.2340e3         notation=sci        
    -12345        -1e4          -1.2e4         -1.23e4         -1.234e4         -1.2345e4         notation=sci        
    -123456       -1e5          -1.2e5         -1.23e5         -1.235e5         -1.2346e5         notation=sci        
    -1234567      -1e6          -1.2e6         -1.23e6         -1.235e6         -1.2346e6         notation=sci        
    -12345678     -1e7          -1.2e7         -1.23e7         -1.235e7         -1.2346e7         notation=sci        
    -123456789    -1e8          -1.2e8         -1.23e8         -1.235e8         -1.2346e8         notation=sci        
    -1.2          -1e0          -1.2e0         -1.20e0         -1.200e0         -1.2000e0         notation=sci        
    -1.23         -1e0          -1.2e0         -1.23e0         -1.230e0         -1.2300e0         notation=sci        
    -1.234        -1e0          -1.2e0         -1.23e0         -1.234e0         -1.2340e0         notation=sci        
    -1.2345       -1e0          -1.2e0         -1.23e0         -1.234e0         -1.2345e0         notation=sci        
    -1.23456      -1e0          -1.2e0         -1.23e0         -1.235e0         -1.2346e0         notation=sci        
    -1.234567     -1e0          -1.2e0         -1.23e0         -1.235e0         -1.2346e0         notation=sci        
    -1.2345678    -1e0          -1.2e0         -1.23e0         -1.235e0         -1.2346e0         notation=sci        
    -1.23456789   -1e0          -1.2e0         -1.23e0         -1.235e0         -1.2346e0         notation=sci        
    -0.000000009  -9e-9         -9.0e-9        -9.00e-9        -9.000e-9        -9.0000e-9        notation=sci        
    -0.000000089  -9e-8         -8.9e-8        -8.90e-8        -8.900e-8        -8.9000e-8        notation=sci        
    -0.000000789  -8e-7         -7.9e-7        -7.89e-7        -7.890e-7        -7.8900e-7        notation=sci        
    -0.000006789  -7e-6         -6.8e-6        -6.79e-6        -6.789e-6        -6.7890e-6        notation=sci        
    -0.000056789  -6e-5         -5.7e-5        -5.68e-5        -5.679e-5        -5.6789e-5        notation=sci        
    -0.000456789  -5e-4         -4.6e-4        -4.57e-4        -4.568e-4        -4.5679e-4        notation=sci        
    -0.003456789  -3e-3         -3.5e-3        -3.46e-3        -3.457e-3        -3.4568e-3        notation=sci        
    -0.023456789  -2e-2         -2.3e-2        -2.35e-2        -2.346e-2        -2.3457e-2        notation=sci        
    -0.123456789  -1e-1         -1.2e-1        -1.23e-1        -1.235e-1        -1.2346e-1        notation=sci

### CODE REVIEW

**Comment by BebeSparkelSparkel** [comment]
Hi Eric,

Thanks for the contributions. Let me review them today. I should be finished tonight.

Best,
Bill

**Comment by BebeSparkelSparkel** [comment]
Issues with to_precison pull request #3

1.) “Should not return numbers with trailing decimals”.
    - Looking at normal conventions a trailing zero should be given if that is the precision prescribed.
    - "A decimal point may be placed after the number; for example "100." indicates specifically that three significant figures are meant"
    - From https://en.wikipedia.org/wiki/Significant_figures#Significant_figures_rules_explained
    - I was unable to have the bar over the top of the zero but having the trailing decimal captures this better than not having it
    - You are right about the 0. being incorrect though so I think that should be kept

2.) I think the api is becoming too complicated for the average use case.
    I seems that std_notation, sci_notation, eng_notation should be the simple wrappers that should be exposed with with input arguments being (value, precision). Then all the more complicated constructors should be put into the to_predision method.
    - Also, a simple auto_notation(value, precision) method should be added that wrapps to_precision with its defaults.

3.) preserve doesn't explain very well what is happening. Perhaps, preserve_integer is a better name.

4.) It isn't a good idea to add 400 rows of numbers to the readme file in the Demonstration section. I'm also having trouble justifying the code in demonstration.py because it doubles the code base with not particularly intuitive code. If we could simplify and shorten this to be < 25 lines I think that it would be far more helpful to users so that they aren't drowning in the amount there.
    - TextGrid should be replaced with https://pypi.python.org/pypi/tabulate
    - there should be more labels in the output to specify the operations being performed
    - format_float seems unnecessary because it seems that to_precision already does this

**Comment by BebeSparkelSparkel** [comment]
I like a your enhancements a lot! Let me know what you think of my review.

**Comment by BebeSparkelSparkel** [comment]
Also, can you please let me know what you are using this for? I think it is good to know what users are actually applying this to

**Comment by epmoyer** [comment]
Thanks William! That was a good review and I agree with almost all of your points:

1.) Agreed! I was unaware of the convention of using the trailing decimal to indicate significance and I thought it was a bug; thanks for pointing me to it. 

ACTION: I will update the PR to restore the old trailing decimal, with the exception of "0."

2.) Agreed; I like that idea.  to_precision calls .std_notation() etc today, so to keep those APIs clean to_precision() would need call the private methods instead, and there'd be a tiny bit of code duplication between the simple wrappers and to_precision, but it would keep the APIs of the wrappers nice and simple so it seems like a good trade off.

ACTION: I'll update the PR to make the simple wrappers all take (value, precision) only, and add an auto_notation() method.

3.) Agreed; I'd started with something longer and chopped it down to make the name short.  preserve_integer isn't bad.  Though if you convert 1234.5 to a precision of 2 you'd get 1200 or 1234 depending on the setting of preserve_integer, and it feels a little funny that there isn't really an integer that you can point to and say you are preserving.  The original number was not an integer, so you're really preserving some digits of the post processed result, which happens to be an integer. Ah the agonies of devising good names :) I'll go with preseve_integer for now, but let's shout if either of us has an epiphany later.

"A beginning is the time for taking the most delicate care that the balances are correct." - Dune, Frank Herbert

4.) 
- I used tabulate originally but found it was modifying the numbers in the table even though they were strings.  I picked thorough the tabulate code just now and there is a "disable_numparse=True" option which wasn't listed on their PyPI page.  I can probably make it work using that (and I agree; I'd rather use tabulate then include the custom class)

- I'm OK with dropping the long table from the README.  If you'd like to keep the module (demonstration.py) around then maybe it would be appropriate to list a small snippet in the README so that people can see what they'll get by running it.  For me I found it was an useful tool when working on the code & for understanding / confirming the behavior of the various options (legacy and new).  My guess it that it will be a handy tool for validating any future changes as well.

- You said "there should be more labels...".  There's a lot of data in the table so I was trying to keep it tight.  Maybe a mix of some initial explanation (printed before the table starts), more descriptive column headings, and potentially more text in the last column could to the trick.  I'll take a crack at it.

- You said "format_float seems unnecessary...".  I originally created demonstration.py as a tool for validating the behavior of to_precision(), so it was important that if there were bugs in to_precision() that they didn't end up corrupting the left hand column.  I still think that's a pretty good strategy, even though I agree one could use to_precision() to generate the same result (if it's working correctly).

ACTIONS:
- I will adopt tabulate
- I will reduce the README to contain a small snippet of the output of demonstration.py, and we'll see if we like that better.
- I will drop demonstration.py's line count & simplify (probably drop the command line args too)
- I will take a crack at making the table more comprehensible to someone who sees it for the first time.

STILL OPEN:
- format_float.  At the very least I can probably do it all on one line without needing the function definition:     '{:0.10f}'.format(value).rstrip('0').rstrip('.')
- You didn't say what you thought about using the set of vectors generated by demostration.py as part of the test suite.  If we did that, we'd run it once, decide we liked the vectors, and commit a test_vectors.json file to the repo, which tests.py would use for validation.  One wouldn't ever regenerate/update the vectors file unless one needed to fix a mistake in the vectors, add vector coverage for new features, or expand the vector coverage for some important case one realized was missing.

----------
As to what I'm using the library for, at work I develop simulations in Jupyter notebooks to explore the predicted outcomes of system architecture choices we are considering in a complex system.  Recently some of the outputs of those simulations take the form of a colored grid (rendered in matplotlib) in which the numbers in the cells are MTBE predictions (Mean Time Between Events, in units of weeks), and the cell colors correspond to "binned" time ranges (weekly, monthly, annually).  The numerical values (the MTBEs) end up spanning several orders of magnitude from 1 down to ~10^-7.  Because the table has a lot of data in it, and is headed to a business/marketing audience, I minimize the visual clutter by rendering the numbers with 3 digits of precision, zero stripping, and auto scientific notation below .00001.

**Comment by epmoyer** [comment]
William,

I've refactored the simple wrapper functions (locally; not pushed yet) to all have arguments of (value, precision) as you suggested, but it occurs to me that the original API for sci_notation and eng_notation also took a filler argument, which was required (it had no default).  

The 3 reasonable API choices are:
- def sci_notation(value, precision):
    - Clean, but breaks with the old API
    - Custom filler can be set by using to_precision()
- def sci_notation(value, precision, filler):
    - Matches old API, but more verbose than one would have probably wished
- def sci_notation(value, precision, filler='e'):
    - New users can leave off the filler and get the default.
    - Old users' code will work unmodified

I think the best choices are the first and the last, depending on whether you are willing to break with the old API.  Which do you prefer?

**Comment by BebeSparkelSparkel** [comment]
Hey Eric,

Hope this makes things more clear.

## format_float
If format float adds a layer of protection then we should keep it.

## Vector Checks
I like the idea of using vectors to check the output, but I would also like it to be more customizable so that people can add/modify specific cases.  I think this could be stored in json or sqlite table with the columns id, raw/value, precision, auto, std, sci, eng, description. json example:\
```json
[
    {"id": 0, "raw": "12", "precision": 2, "auto": "12", "std": "12", "sci": "1.2e1", "eng": "12e0", "description": "optional description of this case"},
    ...
]
```
I think this would be very useful because we could write automatic checks for this and people could fine tune the desired behavior as well. This could also be transformed to a table like yours if that is really what you are looking for.

## More Labels for demonstration.py
I was thinking that just a label for the type being displayed in each section where you have the empty line like "Scientific Notation", "Engineering Notation", etc.

## API Choices
```python
std_notation(value, precision, _)  # change extra to _ because vulture hates extra
auto_notation(value, precision, filler='e')
sci_notation(value, precision, filler='e')
eng_notation(value, precision, filler='e')
```
Also, I'm not sure if filler is the best name, so if you have a better suggestion I'm open to changing it.

## Use Case
Thanks for explaining that.  Yeah, getting data for a business, marketing, or engineering directors (my case) is generally challenging. Do you ever have to deal with error bounds?

**Comment by epmoyer** [comment]
API Choices
------------
How do you like 'delimiter'?
```python
auto_notation(value, precision, delimiter='e')
```

Use Case
---------
I rarely need to deal with error bounds, and never (so far) in any of my coding projects.  It comes up when recording results from lab equipment, but I've not needed to do any numerical analysis on any lab results so far.

**Comment by epmoyer** [comment]
Note
----
* I am de-scoping test vectors from this PR.  We can add them later if we like, but there's a little more discussion to have and it was cleaner to revisit vectors after completing this change.

Changes
-------
* Added auto_notation(value, precision)
* Changed wrapper flow
    * Function previously named _sci_notation() is now _sci_decompose()
    * The following now wrap to_precision():
        * auto_notation(), std_notation(), sci_notation, eng_notation()
    * to_precision() now dispatches to one of:
        * _std_notation(), _sci_notation(), _eng_notation()
* Moved value = float(value) from to_precision() into _number_profile() and _sci_decompose()
    * Now all functions support automatic type conversion of input value
    * Added tests to validate automatic type conversion
* Restored trailing decimal for cases like to_precision(120, 3) => "120." 
* Retained elimination of trailing decimal for single zero, so to_precision(0, 1) => "0" 
* Changed preserve to preseve_integer
* Changed demonstration.py
    * Used tabulate
    * Added section descriptions
    * Print one table per section
    * Summarized function options on one line, before each table
    * Removed "Additional Options" column
    * Simplified test value generation
    * Removed test vector generation
    * Removed all but one negative input value
* tests.py
    * Added test cases for new auto wrapper
* README
    * Added 'Decimal Notation' section
    * Added auto_notation()
    * Updated 'Demonstration' section.  Replaced long table with an excerpt.
    * Updated delimiter argument

**Comment by BebeSparkelSparkel** [comment]
Check out https://github.com/BebeSparkelSparkel/to-precision/tree/epmoyer-enhancements
if you like it I'll pull it.

**Comment by epmoyer** [comment]
I like the readability changes you made to demonstration.py; nice job.
I'm happy to proceed with the pull.

**Comment by BebeSparkelSparkel** [comment]
Hey Eric thanks for your help on improving this a lot. I have just pulled the branch.

Since, you seem to really know this library would you mind being made a colaborator?

**Comment by epmoyer** [comment]
It would be an honor; happy to. Thanks!

**Comment by BebeSparkelSparkel** [comment]
Thank you so much! I'm looking forward to improving this and other projects with you soon.
